---
title: "Transcriptome Analysis in Colorectal Cancer Diagnostics"
author: "Group 2: Joudy Raba Bakri Radwan Mouhaffel, Lea Skaar-Henriksen, Haldor Haugen, Gege Liu"
date: "2024-11-08"
output:
  html_document:
    toc: true            
    toc_depth: 4         
    toc_float: true      
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Abstract

## 1. Introduction
For the BIN315 course project, our group made a decision to investigate a dataset from TCGA-COAD, a Cancer Genome Atlas initiative focusing on colorectal cancer (CRC). Our goal was to apply our knowledge in differential gene expression analysis, machine learning, and enrichment analysis to gain a deeper understanding of this disease.  This report presents our efforts and findings that may provide valuable insights.

### 1.1 Dataset
In this study, we utilized the TCGA-COAD dataset, which is publicly available through the Cancer Genome Atlas (TCGA) research network at https://portal.gdc.cancer.gov/, and contains multi-omics data. We specifically focused on RNA-Seq data, accessed via the TCGAbiolinks R package[1]. This dataset includes a key attribute, sample_type, which allows us to distinguish between primary tumor samples and solid tissue normal samples, making it well-suited for classification analysis.

## 2. Methods
### 2.1 Libraries
The libraries used in our analysis are listed below.
```{r library, echo=FALSE, results='hide'}
suppressMessages({
  # Ensure BiocManager is installed
  if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
  }
  
  # Install required Bioconductor packages if not already installed
  required_packages <- c("TCGAbiolinks", "edgeR", "SummarizedExperiment")
  for (pkg in required_packages) {
    if (!requireNamespace(pkg, quietly = TRUE)) {
      BiocManager::install(pkg)
    }
  }
})
```

```{r library-show, results='hide'}
suppressMessages({
library(TCGAbiolinks)
library(SummarizedExperiment)
library(edgeR)
library(tidyverse)
library(randomForest)
library(matrixStats)
library(clusterProfiler)
library(org.Hs.eg.db)
library(dplyr)
})
```

### 2.2 Exploring and preprocessing
```{r load data from TCGA, echo=FALSE, results='hide'}
# 1. Download RNA-Seq data for selected projects
# sink(tempfile())
# suppressMessages({
# query <- GDCquery(
#   project = "TCGA-COAD",
#   data.category = "Transcriptome Profiling",
#   data.type = "Gene Expression Quantification",
#   workflow.type = "STAR - Counts"
# )
# GDCdownload(query, method = "api", files.per.chunk = 10)
# data <- GDCprepare(query)
# })
# sink()
# 
# # 2. Select the tumor samples and normal samples
# filtered_samples <- colData(data)$sample_type %in% c("Primary Tumor", "Solid Tissue Normal")
# data_filtered <- assays(data)$unstranded[, filtered_samples]
# dim(data_filtered) # 60661 genes 522 samples
# sample_types_filtered <- colData(data)$sample_type[filtered_samples]
# table(sample_types_filtered)  ##Primary Tumor: 481; Solid Tissue Normal:41
# 
# saveRDS(data_filtered, file = "data_filtered.rds")
# saveRDS(sample_types_filtered, file = "sample_types_filtered.rds")
```

#### 2.2.1 Exploring
This dataset contains gene expression data of various samples. It consists of 522 columns (each column represents a different sample) and 60,660 rows (each row corresponds to a gene transcript).

```{r load data}
data_filtered <- readRDS("data_filtered.rds")
sample_types_filtered <- readRDS("sample_types_filtered.rds")
data.counts <- data.frame(data_filtered)
dim(data.counts)
data.counts[1:5,1:2]
```

Then, a sample type distribution plot was drawn, with 481 “primary tumor” type samples and 41 “solid tissue normal” type samples. It can be seen that the number of tumor samples is higher than that of normal tissue samples.

```{r Exploring 1}
names(sample_types_filtered) <- colnames(data.counts)
sample_info <- data.frame(Sample = names(sample_types_filtered), SampleType = sample_types_filtered)

# Plot the distribution
sample_info %>% 
  group_by(SampleType) %>% 
  summarize(Count = n()) %>% 
  ggplot(aes(x = SampleType, y = Count, fill = SampleType)) +
  geom_bar(stat = "identity") +
  theme_bw() +
  labs(title = "Fig.1. Sample Type Distribution", x = "Sample Type", y = "Count") + 
  theme(plot.title = element_text(hjust = 0.5))
```

#### 2.2.2 Preprocessing
We observed that most genes in the dataset have only one transcript, and a few genes (such as ENSG00000002586.20 and ENSG00000002586.20_PAR_Y) have two transcripts. Meanwhile, the ENSG00000002586.20_PAR_Y transcript has zero expression across all samples, providing no meaningful information for downstream analyses. 

```{r Preprocessing-1}
data.counts %>%
  rownames_to_column(var = "Gene") %>% 
  filter(Gene %in% c("ENSG00000002586.20", "ENSG00000002586.20_PAR_Y")) %>% 
  dplyr::select(1:3)
```

Therefore, removing these redundant transcripts can help simplify the dataset and reduce noise while retaining the biological significance of most of the data. After excluding 44 invalid transcripts, we also removed the transcript IDs from the gene names. In addition, considering that the upcoming analysis will involve DESeq2, DESeq2 recommends using safe characters in column names and row names to avoid potential problems. So, we replaced the '.' character in the sample type name with a '-' character.

```{r Preprocessing-2}
data.counts <- data.counts %>% 
  rownames_to_column(var = "Gene") %>%
  filter(!grepl("_PAR", Gene)) %>%
  mutate(Gene = sub("\\..*", "", Gene))

sample_types_filtered <- gsub(" ", "_", sample_types_filtered) 
```

#### 2.2.3 Quality control
From the figure below, we can see that the total reads of most samples are in the range of 2e+07 to 6e+07. The lower reads on the left side of the histogram indicate that that the sequencing depth of some samples may be low, which may mean that the gene expression in these samples is low. To improve the quality and reliability of downstream data analysis, we performed quality control, such as removing low-quality genes and samples.

```{r Filtering-1}
anyNA(data.counts)

data.counts <- data.counts %>% 
  column_to_rownames(var = 'Gene')
sample_totals <- colSums(data.counts)
sample_totals_df <- data.frame(TotalCounts = sample_totals)

ggplot(sample_totals_df, aes(x = TotalCounts)) +
  geom_histogram(bins = 30, fill = "lightblue", color = "grey") +
  labs(title = "Fig.2. Total Reads per Sample", x = "Total Counts", y = "Frequency") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))
```

First, we removed some low-quality samples with less than 20 million expression data. Next, from the gene perspective, we removed genes with low expression in most samples based on the log2 (CPM) transformation of the expression data. After the above steps, the cleaned dataset retained 14,633 genes and 427 samples.

```{r Filtering-2}
# 1. Sample must have 20M reads
idx <- colSums(data.counts) >= 20E6
data.counts.new <- data.counts[,idx]
sample_types_filtered <- sample_types_filtered[idx]
dim(data.counts.new) #60616   427
# 2. Filtering out lowly expressed genes based on the log2(CPM)
data.counts.cpm <- cpm(data.counts.new, log = TRUE)
proportion_threshold <- 0.05
num_samples <- ncol(data.counts.new)
keep_genes <- rowSums(data.counts.cpm > 1) >= (proportion_threshold * num_samples)
data.counts.new = data.counts.new[keep_genes, ]
dim(data.counts.new)
#rm(data.counts.cpm)
```

#### 2.2.4 Normalization 

```{r Normalization-VST}
data_vst <- DESeq2::varianceStabilizingTransformation(as.matrix(data.counts.new))
data_vst <- data_vst - min(data_vst)
```

```{r Normalization-VST2}
# # Plot the expression profile of a random gene after the normalization 
# set.seed(3)
# 
# data_vst %>% 
#   as.data.frame() %>% 
#   rownames_to_column(var = "Gene") %>% 
#   filter(Gene == sample(rownames(data_vst),1)) %>% 
#   pivot_longer(-Gene, names_to = "Sample", values_to = "Expression") %>% 
#   mutate(Sample = as.numeric(row_number())) %>% 
#   ggplot(aes(x = Sample, y = Expression)) +
#   geom_line() +
#   theme_bw()

```


### 2.3 Principal Component Analysis
PCA is an unsupervised dimensionality reduction method that transforms the original feature set into a set of linearly independent variables, i.e., principal components. This transformation can help us not only retain most of the original information when selecting several principle components, but also improve computational efficiency, reduce noise, and enable dataset visualization and feature selection [3]. We tired to performs principal component analysis by the prcomp() function to better understand our dataset and choose the number of principal components for dimensionality reduction.

### 2.4 Differential Expression Analysis 
The purpose of differential expression analysis in this study is to find out the genes whose expression data are upregulated or downregulated in normal samples and cancer samples through statistical tests. This approach helps to provide important insights for cancer research and find the potential biomarkers and mechanisms associated with tumor progression and recurrence [4].

Here, we utilize the DESeq2 package, which is widely-used tool for differential genes analysis of RNA-seq data. DESeq2 package apply the negative binomial models to fit the count-based RNA-Seq data which is different from Limma package that uses linear linear models for the microarray data, as the RNA-Seq data has discrete and over-dispersed characteristics. The key steps in DESeq2 differential expression analysis include data normalization, fitting a negative binomial distribution, estimating dispersion, performing differential expression testing, and correcting for multiple hypothesis testing. Then, the threshold needs to be carefully selected when identifying DEGs (Differentially Expressed Genes), which is an important step to ensure the reliability and accuracy of the analysis results. The choice of measuring statistical significance, such as p-value, p-value adjusted by different techniques (padj), and the size of the difference (expressed as log2 fold change (l2FC)), is crucial because it actually directly affects the reliability of the results and the interpretation of the data [4].

### 2.5 Go Enrichment Analysis


### 2.6 Supervised Methods
Objective: Use supervised learning algorithms to classify samples as tumor or normal based on gene expression data.

#### 2.6.1 Class Balance and Feature selection
#### 2.6.2 Random forest(RF)
#### 2.6.3 K-nearest neighbour(KNN)
#### 2.6.4 Artificial neural networks(ANN)

### 2.7 Evaluation
Objective: Assess the performance of each model using relevant metrics and validation techniques.



## 3. Results

### 3.1 PCA
The results of the PCA are illustrated in Fig. 3 and Fig. 4, which depict the cumulative variance explained by the principal components and the dataset visualization based on PC1 and PC2, respectively.

As can be seen from the table below and Fig.3, the first two components only explain 26.92% of the cumulative variance, which is not very ideal. However, the first 20 principal components can explain 64% of the cumulative variance. 

We can consider the reasons behind the low proportion of variance explained by the first two principal components. First of all,  the dataset includes over 14,000 genes,  which creates a very high-dimensional data space. However, we only have about 400 samples, which is much smaller than the number of genes. This imbanlance makes it harder to capture the variation,  since the dimentionality reductions works better with more samples. Secondly, many genes in the dataset are highly correlated, which adds complexity to the data, resulting each components can explain less variance. Lastly, gene expression data naturally have complex patterns and interactions between genes, making it difficult for the first few principal components to capture a large proportion of the variance.

```{r PCA1}
# Perform the PCA
data.pca <- prcomp(t(data_vst))

# Summary of components
cumulative_variance <- summary(data.pca)$importance[3,]
cumulative_variance[1:30] # show the first 30 principle components
```

Beyond the 60th principal component, the curve of cumulative variance is smoother, indicating that the contributions of other components gradually decrease.

```{r PCA2}
# Plot the cumulative variance explained by the principle components
variance_df <- data.frame(PC = seq_along(cumulative_variance),CumulativeVariance = cumulative_variance)

ggplot(variance_df, aes(x = PC, y = CumulativeVariance)) +
  geom_line(color = "black") +
  geom_point(color = "lightblue") + 
  theme_bw() + 
  labs( title = "Fig.3. Cumulative Variance Ratio", 
        x = "Numbers of Principal Components", 
        y = "Cumulative Variance Explained"
        ) +
  scale_x_continuous(breaks = seq(0, length(cumulative_variance), by = 20)) +
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_text(angle = 45, hjust = 1)
  )
```

In order to interpret our principal components, we hope to find which genes have the greatest impact on each PC. However, the number of original genes is too large, so we focus on the top ten important genes affecting PC1 and PC2.

```{r PCA3}
pc_loadings <- data.pca$rotation

pc_loadings <- pc_loadings %>% 
  as_tibble(rownames = "gene")

genes_pc1 <- pc_loadings %>%
  dplyr::select(gene, PC1) %>%
  pivot_longer(matches("PC"), names_to = "PC", values_to = "loading") %>% 
  group_by(PC) %>% 
  arrange(desc(abs(loading))) %>% 
  slice(1:10) %>% 
  pull(gene) %>% 
  unique()

print(genes_pc1)

genes_pc2 <- pc_loadings %>%
  dplyr::select(gene, PC2) %>%
  pivot_longer(matches("PC"), names_to = "PC", values_to = "loading") %>% 
  group_by(PC) %>% 
  arrange(desc(abs(loading))) %>% 
  slice(1:10) %>% 
  pull(gene) %>% 
  unique()

print(genes_pc2)

```

Then, we turn to visualize our samples on PC1 and PC2 space and distinguish the two classes. In this plot, there is a clear distinction between the primary tumor and solid tissue normal samples. The two classes are mostly separated by PC1, with primary tumors showing a more scattered distribution. And there are a few potential outliers in the lower right of the plot.

```{r PCA4}
classes <- sample_types_filtered %>% factor()

data.pca$x %>% 
  ggplot(aes(PC1, PC2, col = classes)) +
  geom_point() +
  labs(title = "Fig.4. PCA Visualization", x = "PC1", y = "PC2") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5))
```

Therefore, we tried to check the potential outliers found before and remove them, because the number of these outliers was very small, only 8 samples, and these outliers could have a great impact on the subsequent analysis, so we decided to remove them and perform the PCA analysis again to check the results of the before and after PCA.

```{r PCA5}
# Try to find outliers
pca_projection <- data.frame(Sample = rownames(data.pca$x), 
                             PC1 = data.pca$x[, 1], 
                             PC2 = data.pca$x[, 2])

outliers <- pca_projection %>% 
  filter(PC2 < -100)

rownames(outliers)

# Delete the outliers from the dataset
idx_2 <- which(colnames(data_vst) %in% rownames(outliers))

sample_types_filtered_new <- sample_types_filtered[-idx_2]
data_vst <- data_vst %>% 
  as.data.frame() %>% 
  dplyr::select(-rownames(outliers))
```

Now, we can see that the new samples based on PC1 and PC2 are more reasonably distributed, and PC1 seems to play a greater role in splitting the two classes.

```{r PCA6, echo=FALSE}
# Perform the PCA
data.pca <- prcomp(t(data_vst))
classes_new <- sample_types_filtered_new %>% factor()

data.pca$x %>% 
  ggplot(aes(PC1, PC2, col = classes_new)) +
  geom_point() +
  labs(title = "Fig.5. PCA Visualization", x = "PC1", y = "PC2") +
  theme_bw() + 
  theme(plot.title = element_text(hjust = 0.5))
```


### 3.3 Differential Expression Analysis

Here, we create a dataset object and specify the experimental conditions (such as "primary tumor" and "solid tissue normal") though the DESeqDataSetFromMatrix function. Then we use the DESeq function to perform differential expression analysis and extract the analysis results.

```{r DEA1}
# Delete the outliers from the dataset
data.counts.new2 <- data.counts.new %>% 
  as.data.frame() %>% 
  dplyr::select(-rownames(outliers))
dim(data.counts.new2) #17129   419

# Read in RNA-Seq count data and design
suppressMessages({
dds <- DESeq2::DESeqDataSetFromMatrix(
  countData = as.matrix(data.counts.new2),
  colData = data.frame(condition = factor(classes_new, levels = c("Primary_Tumor","Solid_Tissue_Normal"))),
  design = ~ condition)

dds <- DESeq2::DESeq(dds)
res <- DESeq2::results(dds)
as.data.frame(res)[1:10,]
res <- as.data.frame(res)
})
```

Then, we screened out differentially expressed genes based on log2FoldChange and padj, and the thresholds of these two parameters were selected based on the Maurya paper [2]. A total of 2803 genes were screened, including 1389 up-regulated genes and 1414 down-regulated genes.

```{r DEA2}
res0.05 <- res %>% 
  filter(padj< 0.01 & abs(log2FoldChange)>1.5) %>% 
  arrange(padj)

dim(res0.05)

DEGs_up_genes <- res0.05[res0.05$log2FoldChange>0,] %>% 
  rownames_to_column("Gene") %>% 
  dplyr::select("Gene") %>% 
  pull("Gene")

head(res0.05[res0.05$log2FoldChange>0,])
  
DEGs_down_genes <- res0.05[res0.05$log2FoldChange<0,] %>% 
  rownames_to_column("Gene") %>% 
  dplyr::select("Gene") %>% 
  pull("Gene")

head(res0.05[res0.05$log2FoldChange<0,])

#The number of upregulated genes
print(length(DEGs_up_genes))

#The number of downregulated genes
print(length(DEGs_down_genes))
```

Finally, the results can be visualized to easily observe patterns and gain some potential understanding.

```{r DEA3}
res %>% 
  filter(baseMean>0 & log2FoldChange > -10 & !is.na(padj)) %>% 
  mutate(DEGs = ifelse(padj< 0.01 & abs(log2FoldChange)>1.5 , "TRUE", "FALSE")) %>% 
  ggplot(aes(x = log2FoldChange, y = 1/padj, color = DEGs )) +
  geom_point(size = 1) +
  scale_y_log10() +
  labs(x = "log2FoldChange", y = "-log10(padj)", title = "Fig.6. Vulcano plot") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))
```


### 3.3 Go Enrichment Analysis


```{r Enrichmnent Analysis}

go_DEGs_up <- enrichGO(gene = DEGs_up_genes, OrgDb = org.Hs.eg.db, keyType = "ENSEMBL", ont = "BP")
go_DEGs_up_df <- as.data.frame(go_DEGs_up)
#
go_DEGs_up_df[1:6,1:5]

go_DEGs_down <- enrichGO(gene = DEGs_down_genes, OrgDb = org.Hs.eg.db, keyType = "ENSEMBL", ont = "BP", pvalueCutoff = 0.1)
go_DEGs_down_df <- as.data.frame(go_DEGs_down)
#
go_DEGs_down_df[1:6,1:6]

```

### 3.4 Modelling
```{r Modeling}


```



## 4. Discussion

## References
1. Colaprico, A. et al. TCGAbiolinks: an R/Bioconductor package for integrative analysis of TCGA data. Nucleic Acids Res. 44(8), e71 (2016).

2. Maurya, N.S., Kushwaha, S., Chawade, A. et al. Transcriptome profiling by combined machine learning and statistical R analysis identifies TMEM236 as a potential novel diagnostic biomarker for colorectal cancer. Sci Rep 11, 14304 (2021).

3.Ringnér, M. What is principal component analysis?. Nat Biotechnol 26, 303–304 (2008).

4.Rosati, Diletta et al. “Differential gene expression analysis pipelines and bioinformatic tools for the identification of specific biomarkers: A review.” Computational and structural biotechnology journal vol. 23 1154-1168. 1 Mar. 2024, doi:10.1016/j.csbj.2024.02.018





