---
title: "Transcriptome Analysis in Colorectal Cancer Diagnostics"
output: html_document
author: "Group 2: Joudy Raba Bakri Radwan Mouhaffel, Lea Skaar-Henriksen, Haldor Haugen, Gege Liu"
date: "2024-11-08"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Abstract

## 1. Introduction

## 2. Methods

### Dataset
In this study, we utilized the TCGA-COAD dataset, which is publicly available through the Cancer Genome Atlas (TCGA) research network at https://portal.gdc.cancer.gov/, and contains multi-omics data. We specifically focused on RNA-Seq data, accessed via the TCGAbiolinks R package[1]. This dataset includes a key attribute, sample_type, which allows us to distinguish between primary tumor samples and solid tissue normal samples, making it well-suited for classification analysis. The primary objective of this study was to explore how gene expression changes associated with colorectal cancer (CRC) can be identified and used to predict CRC, while applying key concepts in differential analysis, supervised learning methods, and enrichment analysis as taught in the Functional Genomics BIN315 course.

### 2.1 Data Exploration and Preprosessing
#### 2.1.1 Exploration
Plan:
  Load and visualize basic data characteristics, such as the number of samples, genes, and overall data distribution.
Generate basic statistics and visualizations (e.g., distribution plots, correlation matrices) to evaluate data quality initially.

#### 2.1.2 Filtering
Plan:
  Set filtering thresholds (e.g., low expression threshold) to remove genes with low or no expression.

#### 2.1.3 Normalization 
Plan:
  Normalize data using Variance Stabilizing Transformation (VST) to reduce the impact of technical variation.

### 2.2 Principal Component Analysis / Clustering 
Objective: Use PCA to assess whether the data can be classified effectively

Principle: 
  PCA is an unsupervised dimensionality reduction method that transforms the original feature set into a set of linearly independent variables, namely principal components. This transformation can improve computational efficiency while retaining most of the original information, while facilitating visualization exploration and feature selection.

Plan:
  1. Perform PCA and plot the principal components to observe the distribution of tumor and normal samples.
  2. Conduct clustering analysis (e.g., K-means or hierarchical clustering) to further assess the patterns.
  3. Discuss the insights gained from PCA/clustering results and evaluate the feasibility of classification with machine learning.

### 2.3 Differential Expression Analysis 
Objective: Identify genes with significant differential expression between tumor and normal samples.

Plan:
  Calculate log Fold Change (logFC) and False Discovery Rate (FDR) to identify differentially expressed genes.

### 2.4 Supervised Methods
Objective: Use supervised learning algorithms to classify samples as tumor or normal based on gene expression data.

#### 2.4.1 Class Balance and Feature selection
#### 2.4.2 Random forest(RF)
#### 2.4.3 K-nearest neighbour(KNN)
#### 2.4.4 Artificial neural networks(ANN)

### 2.5 Evaluation
Objective: Assess the performance of each model using relevant metrics and validation techniques.

### 2.6 GO enrichment/ Gene set enrichment/ Pathway Analysis
Objective: Investigate the biological significance of differentially expressed genes by identifying enriched biological processes, molecular functions, and pathways.

## 3. Results

### 3.1 Load the dataset 
We start exploring the datasets from the online database or referenced papers.
1. The following code downloads the dataset from an online TCGA database [1].
```{r library, echo=FALSE, results='hide'}
suppressMessages({
if (!requireNamespace("BiocManager", quietly = TRUE)) 
  install.packages("BiocManager")
  BiocManager::install("TCGAbiolinks")
  BiocManager::install("edgeR")
library(TCGAbiolinks)
library(SummarizedExperiment)
# View all projects
# View(TCGAbiolinks:::getGDCprojects())
library(edgeR)
library(tidyverse)
library(randomForest)
library(matrixStats)
})
```

```{r load data from TCGA}
# 1. Download RNA-Seq data for selected projects
sink(tempfile())
suppressMessages({
query <- GDCquery(
  project = "TCGA-COAD",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  workflow.type = "STAR - Counts"
)
GDCdownload(query, method = "api", files.per.chunk = 10)
data <- GDCprepare(query)
})
sink()

# 2. Select the tumor samples and normal samples
## table(colData(data)$sample_type) ##Metastatic:1; Primary Tumor:481; Recurrent Tumor Solid:1; Tissue Normal:41;
filtered_samples <- colData(data)$sample_type %in% c("Primary Tumor", "Solid Tissue Normal")
data_filtered <- assays(data)$unstranded[, filtered_samples]
dim(data_filtered) # 60661 genes 522 samples
sample_types_filtered <- colData(data)$sample_type[filtered_samples]
table(sample_types_filtered)  ##Primary Tumor: 481; Solid Tissue Normal:41

data.counts <- data.frame(data_filtered)
dim(data.counts)
#data.counts$Sample_Type <- sample_types_filtered
```

### 3.2 Exploring and filtering the dataset
The whole datsets have different sequencing depth.
```{r Exploring-1}
#anyNA(data_filtered) check missing data 
sample_totals <- colSums(data_filtered)
hist(sample_totals, breaks = 30, main = "Total Reads per Sample",xlab = "Total Counts")
```

```{r filtering-1}
# 1. Sample must have 20M reads
idx <- colSums(data_filtered) >= 20E6
data_filtered_new <- data_filtered[,idx]
sample_types_filtered_new <- sample_types_filtered[idx]
length(sample_types_filtered_new)
table(sample_types_filtered_new)  ##Primary Tumor：386 Solid Tissue Normal：41
dim(data_filtered_new)
```

```{r filtering-2}
# 2. Filtering out lowly expressed genes based on the log2(CPM)
data_cpm <- cpm(data_filtered_new, log = TRUE)
proportion_threshold <- 0.05
num_samples <- ncol(data_filtered_new)
keep_genes <- rowSums(data_cpm > 1) >= (proportion_threshold * num_samples)
dim(data_filtered_new[keep_genes, ])
data_filtered_new = data_filtered_new[keep_genes, ]

# dynamic parameter
#data_cpm <- cpm(data_filtered_new, log = TRUE)
#qnt.cut <- 0.30
#mean_values <- rowMeans(data_cpm)
#filtered_genes <- mean_values > quantile(mean_values, qnt.cut)
#dim(data_filtered_new[filtered_genes, ])

```

### 3.3 Normalization 
Use DESeq2 to compute VST expression values
```{r Normalization-VST}
data_vst <- DESeq2::varianceStabilizingTransformation(data_filtered_new)
data_vst <- data_vst - min(data_vst)
```


```{r Normalization-VST2}
# Plot the expression profile of a random gene after the normalization 
set.seed(3)

data_vst %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "Gene") %>% 
  filter(Gene == sample(rownames(data_vst),1)) %>% 
  pivot_longer(-Gene, names_to = "Sample", values_to = "Expression") %>% 
  mutate(Sample = as.numeric(row_number())) %>% 
  ggplot(aes(x = Sample, y = Expression)) +
  geom_line() +
  theme_bw()
```

### 3.4 PCA
The first two components only explain 26.92% of the cumulative variance.
```{r PCA1}
data.pca <- prcomp(t(data_vst))
# Summary of components
summary(data.pca)$importance[3,][1:10]
```


```{r PCA2}
classes <- sample_types_filtered_new %>% factor()
data.pca$x %>% 
  ggplot(aes(PC1, PC2, col = classes)) +
  geom_point()
```

### 3.5 Differential Expression Analysis
calculate the log2FC and FDR
```{r DEA1}
# Read in RNA-Seq count data and design
dds <- DESeq2::DESeqDataSetFromMatrix(
  countData = data_filtered_new,
  colData = data.frame(condition = factor(classes, levels = c("Primary Tumor","Solid Tissue Normal"))),
  design = ~ condition)

# Names of samples
colnames(dds)

# Conditions of the samples
dds$condition

dds <- DESeq2::DESeq(dds)
 
```

```{r DEA2}
res <- DESeq2::results(dds)
as.data.frame(res)[1:10,]

res <- as.data.frame(res)

res0.05 <- res %>% 
  filter(padj< 0.05) %>% 
  arrange(padj)

head(res0.05)
#dim(res0.05) #12945

```

2. The following code reads the dataset from the supplementary file S1 provided by the reference paper[2].
```{r load data from paper}
data_from_paper <- read.csv("CRC_paper_dataset.csv", header = TRUE, sep = ",", stringsAsFactors = FALSE)
dim(data_from_paper) # 23182 genes 696 samples
```


## 4. Discussion

## References
1. Colaprico, A. et al. TCGAbiolinks: an R/Bioconductor package for integrative analysis of TCGA data. Nucleic Acids Res. 44(8), e71 (2016).

2. Maurya, N.S., Kushwaha, S., Chawade, A. et al. Transcriptome profiling by combined machine learning and statistical R analysis identifies TMEM236 as a potential novel diagnostic biomarker for colorectal cancer. Sci Rep 11, 14304 (2021).



